### PR TITLE
make compare sam stringent 

### PR DESCRIPTION
**Pull request by akiezun**
Fixes https://github.com/broadinstitute/gatk/issues/1073

note expected.HiSeq.1mb.1RG.2k_lines.bqsr.DIQ.alternate.bam is modified because the previous version was bogus and did not remove indel quals (ie included BD and BI tags, incorrectly)

Note: as discovered by these new stringent tests, cram writing is broken (https://github.com/samtools/htsjdk/issues/364) so we make that comparison lenient for now

Note: Spark tools lose order of reads from the input and so comparison is done after sorting the output files (an alternative would have been to write the stringent sam comparison in a way that can ignore order or equivalent reads, but sorting is easier to do).

### CODE REVIEW

**Comment by akiezun** [comment]
@droazen i'd appreciate a quick review. I need the stringent bam comparisons for my work on bqsr speedup. reassign if appropriate

File: src/main/java/org/broadinstitute/hellbender/utils/Utils.java [hunk 1]
```diff
@@ -441,6 +445,49 @@ public static String calcMD5(final byte[] bytes) {
     }
 
     /**
+     * Calculates the MD5 for the specified file and returns it as a String
+     *
+     * @param file file whose MD5 to calculate
+     * @return file's MD5 in String form
+     * @throws RuntimeException if the file could not be read
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/Utils.java]: `IOException` is not a `RuntimeException`

File: src/main/java/org/broadinstitute/hellbender/utils/Utils.java [hunk 2]
```diff
@@ -441,6 +445,49 @@ public static String calcMD5(final byte[] bytes) {
...
+     */
+    public static String calculateFileMD5( final File file ) throws IOException{
+        return Utils.calcMD5(getBytesFromFile(file));
+    }
+
+    /**
+     * Returns the byte[] of the entire contents of file, for md5 calculations
+     */
+    private static byte[] getBytesFromFile(final File file) throws IOException {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/Utils.java]: Make this method protected and add a direct unit test for it. At a minimum, a test that it reads the correct number of bytes for a particular file (ideally also check the contents).

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/Utils.java]: Actually, I take it back. Please delete this method completely, and replace all usages with calls to `org.apache.commons.io.FileUtils.readFileToByteArray()`

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java [hunk 3]
```diff
@@ -28,11 +30,17 @@
     private final int nOutputFiles;
     private final List<String> expectedFileNames;
 
+    //If this field is set to true, bam files will be compared after they get sorted.
+    //This is needed as a workaround because Spark tools don't respect a pre-ordered BAMs
+    // and so may create BAMs that are sorted differently than the input (though both orders will be valid).
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: Are we ever producing bams in spark that claim to be coordinate-sorted but aren't?

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java [hunk 4]
```diff
@@ -183,8 +196,28 @@ public static void assertEqualTextFiles(final File resultFile, final File expect
...
+            sort1.INPUT = resultFile;
+            sort1.OUTPUT = resultFileSorted;
+            sort1.SORT_ORDER = SAMFileHeader.SortOrder.coordinate;
+            sort1.runTool();
+
+            final SortSam sort2 = new SortSam();
+            sort2.INPUT = expectedFile;
+            sort2.OUTPUT = expectedFileSorted;
+            sort2.SORT_ORDER = SAMFileHeader.SortOrder.coordinate;
+            sort2.runTool();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: Extract a `sortSam(input, output)` helper method, and call it in both places.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 5]
```diff
@@ -27,12 +24,19 @@ private static SamReader getReader(final File sam, final ValidationStringency va
...
+    }
+
+    /**
+     *  causes an exception if the given sam files aren't equal
+     *  @param reference is allowed to be null
+     *  Lenient comparisons only look at the header and alignment information.
+     *  NOTE: Lenient comparisons are a temporary workaround and will be removed in the future
+     */
+    public static void assertSamsEqualLenient(final File sam1, final File sam2, final ValidationStringency validationStringency, final File reference) throws IOException {
+        final String equalStringent = samsEqualLenient(sam1, sam2, validationStringency, reference);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: `equalStringent` is mis-named -- should be named `equalLenient`

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 6]
```diff
@@ -27,12 +24,19 @@ private static SamReader getReader(final File sam, final ValidationStringency va
...
      */
     public static void assertSamsEqual(final File sam1, final File sam2, final ValidationStringency validationStringency, final File reference) throws IOException {
-        try (final SamReader reader1 = getReader(sam1, validationStringency, reference);
-             final SamReader reader2 = getReader(sam2, validationStringency, reference)) {
-            final SamComparison comparison = new SamComparison(reader1, reader2);
-            final boolean equal = comparison.areEqual();
-            Assert.assertTrue(equal, "SAM file " + sam1.getPath() + " differs from expected output:" + sam2.getPath());
-        }
+        final String equalStringent = samsEqualStringent(sam1, sam2, validationStringency, reference);
+        Assert.assertNull(equalStringent, "SAM file " + sam1.getPath() + " differs from expected output:" + sam2.getPath() + " " + equalStringent);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: If `sam2` is always the "expected" output, then it should be named something like `expectedSam`, and `sam1` should be `actualSam`

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 7]
```diff
@@ -44,8 +48,6 @@ public static void assertSamsEqual(final File sam1, final File sam2, final Valid
 
     /**
      * causes an exception if the given sam files aren't equal
-     * @param reference is allowed to be null
-     * the default ValidationStringency value for this method is DEFAULT_STRINGENCY
      */
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Why did you delete this still-valid javadoc? As punishment, you should not only restore it, but add full javadoc for all parameters for public methods in this class.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 8]
```diff
@@ -60,40 +62,12 @@ public static void assertSamsEqual(final File sam1, final File sam2) throws IOEx
...
-     * the default ValidationStringency value for this method is DEFAULT_STRINGENCY
-     */
-    public static void assertSamsNonEqual(final File sam1, final File sam2, final File reference) throws IOException {
-        assertSamsNonEqual(sam1, sam2, ValidationStringency.DEFAULT_STRINGENCY, reference);
-    }
-
-    /**
-     * causes an exception if the given sam files are equal
-     * the default ValidationStringency value for this method is DEFAULT_STRINGENCY
+     * causes an exception if the given sam files aren't equal
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: "aren't equal"? This doc isn't right...

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 9]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
         assertSamValid(sam, ValidationStringency.LENIENT, null);
     }
 
+    /**
+     * Compares SAM/BAM files in a stringent way but not by byte identity (allow reorder of attributes)
+     */
+    private static String samsEqualLenient(File sam1, File sam2, ValidationStringency validation, File reference) throws IOException {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Document your confusing scheme of returning a null `String` from this method when the files are equal.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 10]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            return comparison.areEqual() ? null : "SamComparison fails";
+        }
+    }
+
+    /**
+     * Compares SAM/BAM files in a stringent way but not by byte identity (allow reorder of attributes)
+     * Comparing by MD5s is too strict and comparing by SamComparison is too lenient. So we need this method.
+     * Returns null if equal or message string if not equal.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Returns -> `@return`

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 11]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)) {
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            return comparison.areEqual() ? null : "SamComparison fails";
+        }
+    }
+
+    /**
+     * Compares SAM/BAM files in a stringent way but not by byte identity (allow reorder of attributes)
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Document all of the ways this comparison differs from an exact match.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 12]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+     * Comparing by MD5s is too strict and comparing by SamComparison is too lenient. So we need this method.
+     * Returns null if equal or message string if not equal.
+     */
+    public static String samsEqualStringent(final File sam1, final File sam2, final ValidationStringency validation, final File reference) throws IOException {
+        //Three phases:
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Break this method up into 3 helper methods (1 for each "phase").

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 13]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+
+        //2:  SamComparison says no, then we're not equal
+        try(final SamReader reader1 = getReader(sam1, validation, reference);
+            final SamReader reader2 = getReader(sam2, validation, reference)){
+
+            final SamComparison comparison = new SamComparison(reader1, reader2);
+            if (! comparison.areHeadersEqual()){  //For headers, we want identity (at least for now)
+                return "Headers different";
+            }
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Why require a strict match on the headers? The headers often contain provenance-related info like the tools that have been run on the file with full command lines, and so we should not require that the headers exactly match.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 14]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+            if (!comparison.areEqual()) {
+                return "SamComparison failed";
+            }
+        }
+
+        //At this point we know that the files are not byte-wise identical, but are equal according to SamComparison and their headers are equal
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Why both with the `SamComparison` step? Why not write a `compareHeadersIgnorePGAndCOLines()` method, call that, then go straight into the record-by-record `readsEqualAllowAddingAttributes()` comparison?

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 15]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+            if (it1.hasNext() || it2.hasNext()) {
+                //at least one has no more records (because the while loop is done) and at least one does have records. So we're not equal.
+                return "Not the same number of reads";
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Compares the reads but ignores order or attributes.
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: or -> of

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 16]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+     * Also allows read2 to have a superset of attributes of read1.
+     */
+    private static String readsEqualAllowAddingAttributes(final SAMRecord read1, final SAMRecord read2) {
+        final SAMRecordCoordinateComparator coordinateComparator = new SAMRecordCoordinateComparator();
+        if (!Objects.equals(read1.getReadName(), read2.getReadName())){
+            return "name different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
+        }
+        if (!Objects.equals(read1.getFlags(), read2.getFlags())){
+            return "flags different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Include the differing flag values in the error message

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 17]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+        if (!Objects.equals(read1.getInferredInsertSize(), read2.getInferredInsertSize())){
+            return "getInferredInsertSize different read1:" + read1.getReadName() + " read2:" + read2.getReadName() + " (" + read1.getInferredInsertSize()+ " vs " + read2.getInferredInsertSize() + ")";
+        }
+        if (!Objects.equals(read1.getMappingQuality(), read2.getMappingQuality())){
+            return "getMappingQuality different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Include the differing mapping quality values in the error message.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 18]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+        //Note the comparator does check flags, MQ and insert size but we want better messages
+        if (0 != coordinateComparator.fileOrderCompare(read1, read2)){
+            return "file order read1:" + read1.getContig() + ":" + read1.getAlignmentStart() + " read2:" + read2.getContig() + ":" + read2.getAlignmentStart();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Don't use a `SAMRecordCoordinateComparator` here -- it potentially duplicates a lot of comparisons. Instead, just check directly that the reference index/name and alignment start are equal, and adds checks here for the mate position as well.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 19]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+        if (!Objects.equals(read1.getCigar(), read2.getCigar())){
+            return "getCigar different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Include the differing Cigars in the error message.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 20]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+        if (!Arrays.equals(read1.getReadBases(), read2.getReadBases())){
+            return "getReadBases different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Include the differing base arrays in the error message.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 21]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+        if (!Arrays.equals(read1.getBaseQualities(), read2.getBaseQualities())){
+            return "getBaseQualities different read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Include the differing base quality arrays in the error message.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 22]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+        final List<SAMRecord.SAMTagAndValue> attr1 = read1.getAttributes();
+        final List<SAMRecord.SAMTagAndValue> attr2 = read2.getAttributes();
+        if (attr1.size() > attr2.size()){
+            return "lost attributes read1:" + read1.getReadName() + " read2:" + read2.getReadName() + "("+ attr1.size() + " vs " + attr2.size() +")" ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: This message could be worded more clearly -- "read1 contains attributes that read2 lacks"

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 23]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        }
+
+        //We want to compare attributes regardless of order, so we sort first attributes by name
+        final SortedMap<String, Object> values1 = new TreeMap<>();
+        final SortedMap<String, Object> values2 = new TreeMap<>();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: These are poorly named -- `attributeValues`

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Actually, why do you need a `SortedMap` here? You are never iterating over the Map contents, just calling `containsKey()` and `get()`. A `HashMap` would do the same job without the cost of the sort.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 24]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+        for (SAMRecord.SAMTagAndValue samTagAndValue : attr2) {
+            values2.put(samTagAndValue.tag, samTagAndValue.value);
+        }
+
+        for (int i = 0; i < attr1.size(); i++) {
+            final String tag = attr1.get(i).tag;
+            if (!values2.containsKey(tag)){
+                return "attribute " + tag + " is missing from read2:" + read2.getReadName() ;
+            }
+            final Object v1 = values1.get(tag);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: You could already have this Object ready if you saved the return value of `attr1.get(i)` above (avoiding an extra hash lookup per read1 attribute).

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 25]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+            final Object v2 = values2.get(tag);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Instead of calling both `containsKey()` and `get()` on `values2`, just call `get()` and check the return value for null (saves one additional hash lookup per attribute).

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 26]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+            if (!Objects.equals(v1, v2)){
+                return "attribute " + tag + " has different value in read1:" + read1.getReadName() + " read2:" + read2.getReadName() ;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Include the differing attribute values in the error message.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java [hunk 27]
```diff
@@ -135,4 +109,126 @@ public static void assertSamValid(final File sam) throws IOException {
...
+            }
+            return null;
+        }
+    }
+
+    /**
+     * Compares the reads but ignores order or attributes.
+     * Also allows read2 to have a superset of attributes of read1.
+     */
+    private static String readsEqualAllowAddingAttributes(final SAMRecord read1, final SAMRecord read2) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: Extract the attribute comparison code out as a separate helper method to break this long-ish method up.

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 28]
```diff
@@ -44,7 +44,12 @@ public void testFileToFileCRAM(String fileIn, String extOut, String reference) t
                 "-R", refFile.getAbsolutePath()
         };
         runCommandLine(args);
-        SamAssertionUtils.assertSamsEqual(ORIG_BAM, outFile, refFile);
+        //TODO: Make this comparison non-lenient in all cases. Waits on https://github.com/samtools/htsjdk/issues/364
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: You need to create a hellbender ticket for this TODO, otherwise we will forget about while waiting for https://github.com/samtools/htsjdk/issues/364 to be fixed.

File: src/test/java/org/broadinstitute/hellbender/tools/picard/sam/GatherBamFilesIntegrationTest.java [hunk 29]
```diff
@@ -40,6 +57,21 @@ public void testTheGathering() throws Exception {
         args.add(outputFile.getAbsolutePath());
         runCommandLine(args);
         SamAssertionUtils.assertSamsEqual(ORIG_BAM, outputFile);
-        SamAssertionUtils.assertSamsNonEqual(ORIG_BAM, SPLIT_BAMS.get(0)); // sanity check
+        Assert.assertNotNull(SamAssertionUtils.samsEqualStringent(ORIG_BAM, SPLIT_BAMS_UNMAPPED_LAST.get(0), ValidationStringency.DEFAULT_STRINGENCY, null)); // sanity check
     }
+
+    @Test
+    public void testTheGatheringUnmappedFirst() throws Exception {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/sam/GatherBamFilesIntegrationTest.java]: Use a `DataProvider` to provide the `SPLIT_BAMS_UNMAPPED_FIRST` and `SPLIT_BAMS_UNMAPPED_LAST` lists, and combine this test with the previous test.

File: src/test/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtilsUnitTest.java [hunk 30]
```diff
@@ -0,0 +1,44 @@
+package org.broadinstitute.hellbender.utils.test;
+
+import htsjdk.samtools.ValidationStringency;
+import org.junit.Assert;
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtilsUnitTest.java]: Wrong `Assert` class -- use TestNG's `Assert` instead

File: src/test/resources/org/broadinstitute/hellbender/utils/test/SamAssertionUtilsUnitTest/file1.sam [hunk 31]
```diff
@@ -0,0 +1,25 @@
+@HD	VN:1.5	GO:none	SO:coordinate
```
Comment (droazen) [inline src/test/resources/org/broadinstitute/hellbender/utils/test/SamAssertionUtilsUnitTest/file1.sam]: What is the new `expected.HiSeq.1mb.1RG.2k_lines.bqsr.DIQ.alternate.bam` for? It does not appear to be used, or to be related to this PR at all.

**Comment by droazen** [comment]
Review complete -- back to @akiezun

File: src/main/java/org/broadinstitute/hellbender/utils/Utils.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/Utils.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/utils/Utils.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/Utils.java]: fixed. i forgot about that method

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: I dont think so. And SortSam would blow up if that's the case.

File: src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/IntegrationTestSpec.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 7 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: fixed for all `equals` methods

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 8 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 9 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 10 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 11 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 13 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done.

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 14 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 15 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 16 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 17 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 18 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 19 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 20 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 21 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 22 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 23 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 24 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 25 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 26 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java (hunk 27 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtils.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java (hunk 28 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: done https://github.com/broadinstitute/gatk/issues/1087

File: src/test/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtilsUnitTest.java (hunk 30 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/utils/test/SamAssertionUtilsUnitTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/picard/sam/GatherBamFilesIntegrationTest.java (hunk 29 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/sam/GatherBamFilesIntegrationTest.java]: done

**Comment by akiezun** [comment]
back to @droazen - all comments addressed. It's in 3 parts.

**Comment by droazen** [comment]
Second-pass review complete. Rebase/merge when done addressing remaining comments.
